c++/src/kj/common.h

```diff
@@ -1748,6 +1749,9 @@ class ArrayPtr: public DisallowConstCopyIfNotConst<T> {
     KJ_IREQUIRE(start <= end && end <= size_, "Out-of-bounds ArrayPtr::slice().");
     return ArrayPtr(ptr + start, end - start);
   }
+  inline bool startsWith(const ArrayPtr<const T>& other) const {
+    return other.size_ <= size_ && memcmp(ptr, other.ptr, other.size_) == 0;
```

Comment:
`memcmp()` is only correct here if `T` is a basic integral type.

Please do:

```
  inline bool startsWith(const ArrayPtr<const T>& other) const {
    return other.size_ <= size_ && slice(0, other.size_) == other;
  }
```

We already have an optimization so that `operator==` will use `memcmp()` when `T` is integral.

Reply:
https://github.com/capnproto/capnproto/pull/1480/commits/3331537fc308364e6062575a5270b35aad53dc89